File: BittraBLETest/BLE/BLEPeripheral.swift
import Foundation
import CoreBluetooth

final class BLEPeripheral: NSObject, CBPeripheralManagerDelegate {

  private var pm: CBPeripheralManager!
  private let log: (String, String) -> Void

  private var reqChar: CBMutableCharacteristic!
  private var chunkChar: CBMutableCharacteristic!
  private var service: CBMutableService!

  private var myTeaser: String = ""
  private var myBody: String = ""

  // 送信キュー
  private var pendingNotifyPackets: [Data] = []
  private var isNotifying: Bool = false

  init(log: @escaping (String, String) -> Void) {
    self.log = log
    super.init()
    self.pm = CBPeripheralManager(delegate: self, queue: nil)
  }

  func setContent(teaser: String, body: String) {
    self.myTeaser = teaser
    self.myBody = body
  }

  func start() {
    guard pm.state == .poweredOn else {
      log("PERIPH", "Bluetooth not poweredOn yet")
      return
    }
    setupGATTIfNeeded()
    startAdvertising()
  }

  func stop() {
    pm.stopAdvertising()
    log("PERIPH", "stopAdvertising")
  }

  private func setupGATTIfNeeded() {
    if service != nil { return }

    reqChar = CBMutableCharacteristic(
      type: GATTProfile.reqCharUUID,
      properties: [.writeWithoutResponse],
      value: nil,
      permissions: [.writeable]
    )

    chunkChar = CBMutableCharacteristic(
      type: GATTProfile.chunkCharUUID,
      properties: [.notify],
      value: nil,
      permissions: []
    )

    service = CBMutableService(type: GATTProfile.serviceUUID, primary: true)
    service.characteristics = [reqChar, chunkChar]

    pm.removeAllServices()
    pm.add(service)
    log("PERIPH", "GATT service added")
  }

  private func startAdvertising() {
    let nonce = UInt16.random(in: 0...UInt16.max)
    let payload = PayloadCodec.encodeAdvert(teaser: myTeaser, nonce: nonce)

    pm.startAdvertising([
      CBAdvertisementDataServiceUUIDsKey: [GATTProfile.serviceUUID],
      CBAdvertisementDataServiceDataKey: [GATTProfile.serviceUUID: payload]
    ])
    log("ADV", "startAdvertising teaser=\(PayloadCodec.normalizeTeaser(myTeaser)) bytes=\(payload.count)")
  }

  // MARK: - CBPeripheralManagerDelegate

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    log("PERIPH", "state=\(peripheral.state.rawValue)")
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    for r in requests {
      guard r.characteristic.uuid == GATTProfile.reqCharUUID else { continue }
      // REQ受信＝即送信
      log("GATT", "REQ received from central")
      // 返信はNotifyなので respond は不要（WriteWithoutResponse）
      sendBodyPreviewThenRest()
    }
  }

  func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
    // 送信バッファが空いたので続行
    flushNotifyQueue()
  }

  // MARK: - Body sending

  private func sendBodyPreviewThenRest() {
    pendingNotifyPackets.removeAll()
    isNotifying = false

    let bytes = Data(myBody.utf8)
    let preview = bytes.prefix(GATTProfile.previewBytes)
    let rest = bytes.dropFirst(GATTProfile.previewBytes)

    // チャンク化（簡易：20B近辺で切る。実際はMTUで調整してOK）
    // iOSはupdateValueが返すfalseで詰まりが見えるので、ここは小さめで安全運転
    let chunkPayloadSize = 140 // 多少大きくてもOK。失敗したらready callbackで再送。
    var seq: UInt16 = 0

    func enqueueChunks(_ data: Data, isLastBlockPotentially: Bool) {
      var offset = 0
      while offset < data.count {
        let end = min(data.count, offset + chunkPayloadSize)
        let slice = data.subdata(in: offset..<end)
        offset = end

        var packet = Data()
        packet.append(UInt8(seq & 0xFF))
        packet.append(UInt8((seq >> 8) & 0xFF))
        packet.append(UInt8(0)) // flags placeholder
        packet.append(slice)
        pendingNotifyPackets.append(packet)
        seq += 1
      }
    }

    enqueueChunks(Data(preview), isLastBlockPotentially: rest.isEmpty)
    enqueueChunks(Data(rest), isLastBlockPotentially: true)

    // lastフラグを最後のpacketに立てる
    if !pendingNotifyPackets.isEmpty {
      pendingNotifyPackets[pendingNotifyPackets.count - 1][2] = 0x01
    }

    log("GATT", "enqueue notify packets=\(pendingNotifyPackets.count) bodyBytes=\(bytes.count)")
    flushNotifyQueue()
  }

  private func flushNotifyQueue() {
    guard !isNotifying else { return }
    isNotifying = true

    while !pendingNotifyPackets.isEmpty {
      let packet = pendingNotifyPackets[0]
      let ok = pm.updateValue(packet, for: chunkChar, onSubscribedCentrals: nil)
      if !ok {
        // バッファ詰まり：ready callbackで再開
        log("GATT", "updateValue blocked; wait ready")
        isNotifying = false
        return
      }
      pendingNotifyPackets.removeFirst()
    }
    log("GATT", "notify complete")
    isNotifying = false
  }
}

File: android/app/src/main/kotlin/com/kaeruko/bittora/ble/BlePeripheral.kt
package com.kaeruko.bittora.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.util.Log
import java.util.UUID

@SuppressLint("MissingPermission")
class BlePeripheral(private val context: Context, private val log: (String, String) -> Unit) {

    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val adapter: BluetoothAdapter? = bluetoothManager.adapter
    private val advertiser = adapter?.bluetoothLeAdvertiser

    private var gattServer: BluetoothGattServer? = null
    private var isAdvertising = false

    private var myTeaser: String = ""
    private var myBody: String = ""

    private val pendingNotifyPackets = mutableListOf<ByteArray>()
    private var isNotifying = false
    private var subscribedDevice: BluetoothDevice? = null

    private val mainHandler = Handler(Looper.getMainLooper())

    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
            log("ADV", "startAdvertising success")
            isAdvertising = true
        }

        override fun onStartFailure(errorCode: Int) {
            log("ADV", "startAdvertising fail: $errorCode")
            isAdvertising = false
        }
    }

    private val gattServerCallback = object : BluetoothGattServerCallback() {
        override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
            log("PERIPH", "connectionStateChange ${device.address} status=$status newState=$newState")
        }

        override fun onCharacteristicWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            characteristic: BluetoothGattCharacteristic,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            if (characteristic.uuid == GATTProfile.REQ_CHAR_UUID) {
                log("GATT", "REQ received from central")
                
                // Write without response
                if (responseNeeded) {
                    gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
                }
                
                sendBodyPreviewThenRest(device)
            }
        }

        override fun onDescriptorWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            descriptor: BluetoothGattDescriptor,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            if (BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE.contentEquals(value)) {
                subscribedDevice = device
            }
            if (responseNeeded) {
                gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
            }
        }

        override fun onNotificationSent(device: BluetoothDevice, status: Int) {
            super.onNotificationSent(device, status)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                isNotifying = false
                flushNotifyQueue(device)
            } else {
                log("GATT", "onNotificationSent error $status")
                isNotifying = false
            }
        }
    }

    fun setContent(teaser: String, body: String) {
        this.myTeaser = teaser
        this.myBody = body
    }

    fun start() {
        if (adapter == null || !adapter.isEnabled) {
            log("PERIPH", "Bluetooth not poweredOn yet")
            return
        }

        // Stop existing advertisement before restarting with new content
        if (isAdvertising) {
            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
        }

        setupGATTIfNeeded()
        startAdvertising()
    }

    fun stop() {
        if (isAdvertising) {
            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
            log("PERIPH", "stopAdvertising")
        }
    }

    private fun setupGATTIfNeeded() {
        if (gattServer != null) return

        gattServer = bluetoothManager.openGattServer(context, gattServerCallback)

        val service = BluetoothGattService(GATTProfile.SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)

        val reqChar = BluetoothGattCharacteristic(
            GATTProfile.REQ_CHAR_UUID,
            BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE or BluetoothGattCharacteristic.PROPERTY_WRITE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )

        val chunkChar = BluetoothGattCharacteristic(
            GATTProfile.CHUNK_CHAR_UUID,
            BluetoothGattCharacteristic.PROPERTY_NOTIFY,
            BluetoothGattCharacteristic.PERMISSION_READ
        )

        val cccDescriptor = BluetoothGattDescriptor(
            UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"),
            BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE
        )
        chunkChar.addDescriptor(cccDescriptor)

        service.addCharacteristic(reqChar)
        service.addCharacteristic(chunkChar)

        gattServer?.addService(service)
        log("PERIPH", "GATT service added")
    }

    private fun startAdvertising() {
        if (advertiser == null) return

        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
            .setConnectable(true)
            .setTimeout(0)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
            .build()

        val nonce = (0..65535).random()
        val payload = PayloadCodec.encodeAdvert(myTeaser, nonce)

        // Main packet: service UUID only (for iOS scan filter)
        val advertiseData = AdvertiseData.Builder()
            .addServiceUuid(ParcelUuid(GATTProfile.SERVICE_UUID))
            .setIncludeDeviceName(false)
            .build()

        // Scan response: teaser payload in manufacturer data (4 bytes overhead vs 18 for service data)
        val scanResponse = AdvertiseData.Builder()
            .addManufacturerData(GATTProfile.MANUFACTURER_ID, payload)
            .setIncludeDeviceName(false)
            .build()

        advertiser.startAdvertising(settings, advertiseData, scanResponse, advertiseCallback)
        log("ADV", "startAdvertising teaser=${PayloadCodec.normalizeTeaser(myTeaser)} bytes=${payload.size}")
    }

    private fun sendBodyPreviewThenRest(device: BluetoothDevice) {
        pendingNotifyPackets.clear()
        isNotifying = false

        val bytes = myBody.toByteArray(Charsets.UTF_8)
        val preview = if (bytes.size > GATTProfile.PREVIEW_BYTES) bytes.copyOfRange(0, GATTProfile.PREVIEW_BYTES) else bytes
        val rest = if (bytes.size > GATTProfile.PREVIEW_BYTES) bytes.copyOfRange(GATTProfile.PREVIEW_BYTES, bytes.size) else ByteArray(0)

        // MTU usually 23 bytes minimum, meaning 20 bytes payload. We use slightly larger if MTU is negotiated, but stick to 20 for safety on Android.
        val chunkPayloadSize = 20
        var seq = 0

        fun enqueueChunks(data: ByteArray, isLastBlockPotentially: Boolean) {
            var offset = 0
            while (offset < data.size) {
                val end = Math.min(data.size, offset + chunkPayloadSize)
                val slice = data.copyOfRange(offset, end)
                offset = end

                val packet = ByteArray(3 + slice.size)
                packet[0] = (seq and 0xFF).toByte()
                packet[1] = ((seq shr 8) and 0xFF).toByte()
                packet[2] = 0 // flags

                System.arraycopy(slice, 0, packet, 3, slice.size)
                pendingNotifyPackets.add(packet)
                seq++
            }
        }

        enqueueChunks(preview, rest.isEmpty())
        enqueueChunks(rest, true)

        if (pendingNotifyPackets.isNotEmpty()) {
            val lastIndex = pendingNotifyPackets.size - 1
            val lastPacket = pendingNotifyPackets[lastIndex]
            lastPacket[2] = 0x01
            pendingNotifyPackets[lastIndex] = lastPacket
        }

        log("GATT", "enqueue notify packets=${pendingNotifyPackets.size} bodyBytes=${bytes.size}")
        flushNotifyQueue(device)
    }

    private fun flushNotifyQueue(device: BluetoothDevice) {
        if (isNotifying || pendingNotifyPackets.isEmpty()) return
        
        isNotifying = true
        val packet = pendingNotifyPackets.removeAt(0)

        val service = gattServer?.getService(GATTProfile.SERVICE_UUID)
        val char = service?.getCharacteristic(GATTProfile.CHUNK_CHAR_UUID)
        
        if (char != null) {
            char.value = packet
            val success = gattServer?.notifyCharacteristicChanged(device, char, false) ?: false
            if (!success) {
                log("GATT", "notifyCharacteristicChanged failed, buffer might be full")
                // On Android, if it returns false we should ideally retry after a short delay
                pendingNotifyPackets.add(0, packet) // put it back
                isNotifying = false
                mainHandler.postDelayed({ flushNotifyQueue(device) }, 50)
            }
        } else {
            isNotifying = false
        }
    }
}
